### PR TITLE
Update the test runner to use the updated code in astropy 1.3

### PR DESCRIPTION
**Pull request by Cadair**
This modifies our internal test runner (the thing that provides `sunpy.self_test` and `python setup.py test`) to be much closer to astropy's implementation. The differences between us and astropy now are the replacement of the astropy `remote_data` argument with `online` and `offline` and the addition of the `figure` argument.

This means we have many more options to `self_test` and `setup.py test` such as checking for open file handles, running pep8 checks etc.

### CODE REVIEW

**Comment by Cadair** [comment]
there is a lot more todo on this, for some reason the doctests are not being picked up by pytest. Also I am working on porting the extra plugins for pytest that astropy uses to provide things like open file checking into a form where we can use them (see astropy/astropy#5770). Oh and it just fails, so that needs fixing as well.

**Comment by Cadair** [comment]
I will close this for now, there may be a couple of different ways of achieving it, like using the new code I wrote in Astropy to expose our current pytest UI.

**Comment by Cadair** [comment]
This should now be resurrected as since this was closed we have adopted doctestplus and remote_data and all the astropy pytest plugins have been broken out into seperate packages, which should make this whole thing a lot easier.

I think the largest challenge here is now the setup command, but it should be doable.

File: sunpy/__init__.py [hunk 1]
```diff
@@ -23,7 +23,9 @@
 
 from sunpy.util.config import load_config, print_config
 from sunpy.util import system_info
-from sunpy.tests import main as self_test
+from sunpy.tests.runner import SunPyTestRunner
+
+self_test = SunPyTestRunner.make_test_runner_in(__path__[0])
```
Comment (Cadair) [inline sunpy/__init__.py]: This needs to not be defined if `ASTROPY_SETUP` is true.

File: sunpy/tests/runner.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            return ['-k-online']
+        elif self.is_online and not offline:
+            return ['-k', 'online']
+        elif not self.is_online and not offline:
+            raise ValueError("offline and online can not both be False.")
+        else:
+            return []
+
+    @keyword(False)
+    def figure(self, figure, kwargs):
```
Comment (Cadair) [inline sunpy/tests/runner.py]: need to add a `figure_only`

File: sunpy/tests/setup_command.py [hunk 3]
```diff
@@ -96,35 +54,43 @@ def generate_testing_command(self):
...
         offline = not self.online_only
         figure = self.figure
         figure_only = self.figure_only
 
+        if self.coverage:
+            pre, post = self._generate_coverage_commands()
+            cmd_pre += pre
+            cmd_post += post
+
         cmd = ('{cmd_pre}{0}; import {1.package_name}, sys; result = ('
```
Comment (Cadair) [inline sunpy/tests/setup_command.py]: need to make sure this list is upto date with the one in astropy.

**Comment by Cadair** [comment]
Online test fails looks to be because of a dead NOAA server.

@sunpy/sunpy-maintainers review please!

File: .gitignore [hunk 4]
```diff
@@ -75,3 +75,4 @@ $RECYCLE.BIN/
 .hypothesis/
 .pytest_cache/
 sunpydata.sqlite
+v/
```
Comment (nabobalis) [inline .gitignore]: What is this?

File: .travis.yml [hunk 5]
```diff
@@ -58,20 +58,12 @@ matrix:
          # Order of tests run in build stages
          - python: 3.6
            stage: Initial tests
-           env: SETUP_CMD='test --coverage'
```
Comment (nabobalis) [inline .travis.yml]: So what does this run now?

File: docs/guide/acquiring_data/database.rst [hunk 6]
```diff
@@ -480,13 +480,13 @@ year:
     >>> for database_entry in database:
     ...     if database_entry.observation_time_start.month in [3,4,5]:
     ...         database.tag(database_entry, 'spring')
-    >>> database.tags
+    >>> database.tags  # doctest: +SKIP
```
Comment (nabobalis) [inline docs/guide/acquiring_data/database.rst]: Why were these tests skipped?

File: docs/guide/data_types/timeseries.rst [hunk 7]
```diff
@@ -34,6 +34,8 @@ following into your interactive Python shell: ::
     >>> import sunpy.data.sample  # doctest: +REMOTE_DATA
     >>> my_timeseries = ts.TimeSeries(sunpy.data.sample.GOES_XRS_TIMESERIES, source='XRS')  # doctest: +REMOTE_DATA
 
+.. doctest-skip-all
```
Comment (nabobalis) [inline docs/guide/data_types/timeseries.rst]: But why? :(

File: sunpy/net/helio/hec.py [hunk 8]
```diff
@@ -163,7 +159,7 @@ def time_query(self, start_time, end_time, table=None, max_records=None):
         >>> hc = hec.HECClient()  # doctest: +REMOTE_DATA
         >>> start = '2005/01/03'
         >>> end = '2005/12/03'
-        >>> temp = hc.time_query(start, end, max_records=10)   # doctest: +REMOTE_DATA
+        >>> temp = hc.time_query(start, end, max_records=10)   # doctest: +REMOTE_DATA +SKIP
```
Comment (nabobalis) [inline sunpy/net/helio/hec.py]: Skip?! :(

File: sunpy/net/helio/hec.py [hunk 9]
```diff
@@ -223,7 +219,7 @@ def make_table_list(self):
         --------
         >>> from sunpy.net.helio import hec
         >>> hc = hec.HECClient()  # doctest: +REMOTE_DATA
-        >>> hc.make_table_list()  # doctest: +REMOTE_DATA
+        >>> hc.make_table_list()  # doctest: +REMOTE_DATA +SKIP
```
Comment (nabobalis) [inline sunpy/net/helio/hec.py]: But my skip?!

File: .travis.yml (hunk 5 above)
Comment (Cadair) [reply .travis.yml]: the same thing, it just uses the default value

File: .gitignore (hunk 4 above)
Comment (Cadair) [reply .gitignore]: :woman_shrugging:

File: docs/guide/data_types/timeseries.rst (hunk 7 above)
Comment (Cadair) [reply docs/guide/data_types/timeseries.rst]: because they don't work!!

File: .travis.yml (hunk 5 above)
Comment (nabobalis) [reply .travis.yml]: Oh that was the default value already. Nevermind.

**Comment by Cadair** [comment]
oh balls: https://travis-ci.org/sunpy/sunpy/jobs/378357403#L2557

The `-k remote_data` in the online build skips all the damn doctests

File: sunpy/net/helio/hec.py (hunk 9 above)
Comment (Cadair) [reply sunpy/net/helio/hec.py]: this one is real. It's just nonsense.

Comment (Cadair) [reply sunpy/net/helio/hec.py]: oh that's it, this function needs user input, so it can't be run as a doctest.

File: sunpy/net/helio/hec.py (hunk 8 above)
Comment (Cadair) [reply sunpy/net/helio/hec.py]: this one also needs user input.

**Comment by Cadair** [comment]
Assuming my rebase hasn't broken anything this is now good to go. Reviews please!

**Pull request merged**